### PR TITLE
Add CONCAT_WS function

### PR DESCRIPTION
**Pull request by yamanogluberk**
Adds support for `CONCAT_WS` function, which takes variable sized `VARCHAR`(min 3) and returns a `VARCHAR` that consists of the concatenation of the arguments except the first one using the first argument as a separator. 

Example query: 
```
SELECT CONCAT_WS('-', 'a', 'b' ) FROM TABLE (generate_series(1,3))
```

Example Result: 
```
a-b               
a-b                 
a-b 
```

### CODE REVIEW

**Comment by yamanogluberk** [comment]
Tests are not implemented yet

**Review by hakanmemisoglu — CHANGES_REQUESTED** [review]

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        ConcatWSFunction function = ConcatWSFunction.create(ConstantExpression.create("-", VARCHAR), ConstantExpression.create("1", VARCHAR), ConstantExpression.create("2", VARCHAR));
+
+        checkEquals(function, ConcatWSFunction.create(ConstantExpression.create("-", VARCHAR), ConstantExpression.create("1", VARCHAR), ConstantExpression.create("2", VARCHAR)), true);
+        checkEquals(function, ConcatWSFunction.create(ConstantExpression.create("-", VARCHAR), ConstantExpression.create("1", VARCHAR), ConstantExpression.create("20", VARCHAR)), false);
+        checkEquals(function, ConcatWSFunction.create(ConstantExpression.create("-", VARCHAR), ConstantExpression.create("2", VARCHAR)), false);
+        checkEquals(function, ConcatWSFunction.create(ConstantExpression.create("1", VARCHAR), ConstantExpression.create("2", VARCHAR)), false);
+    }
+
+    @Test
+    public void testSerialization() {
```
[the remark this answers is not in the document]
Comment (yamanogluberk) [reply hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java:162]: Can you be more specific? I have already tested some cases where it the query should fail.

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    @Test
+    public void testColumn() {
+        for (ExpressionType<?> type1 : ExpressionTypes.all()) {
+            // Test supported types
+            for (ExpressionType<?> type2 : ExpressionTypes.all()) {
+                Class<? extends ExpressionBiValue> clazz = ExpressionBiValue.createBiClass(type1.typeName(), type2.typeName());
+
+                checkColumns(
+                        ExpressionBiValue.createBiValue(clazz, 0, type1.valueFrom(), type2.valueFrom()),
+                        "" + type1.valueFrom() + "-" + type2.valueFrom()
```
[the remark this answers is not in the document]
Comment (yamanogluberk) [reply hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java:54]: Fixed

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+    private void check(String operands, String expectedResult, Object... params) {
+        String sql = "SELECT " + operands + " FROM map";
+
+        checkValue0(sql, SqlColumnType.VARCHAR, expectedResult, params);
+    }
+
+    private void checkFail(String sql, Object... params) {
+        try {
+            List<SqlRow> rows = execute(member, sql, params);
```
[the remark this answers is not in the document]
Comment (yamanogluberk) [reply hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java:183]: Fixed

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (hakanmemisoglu) [reply hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java:162]: There is only `checkEquals()` in ` testSerialization()`. I think you should also add `checkNotEquals()` scenario.

[the remark this answers is not in the document]
Comment (yamanogluberk) [reply hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java:164]: Added

File: hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastConcatWSOperator.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    @Override
+    public boolean checkOperandTypes(HazelcastCallBinding binding, boolean throwOnFailure) {
+        HazelcastSqlValidator validator = binding.getValidator();
+
+
+        if (binding.getOperandType(0).getSqlTypeName() != VARCHAR) {
+            if (throwOnFailure) {
+                throw new AssertionError("");
+            }
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/string/HazelcastConcatWSOperator.java:67]:
```suggestion
        if (binding.getOperandType(0).getSqlTypeName() != VARCHAR) {
            if (throwOnFailure) {
                throw binding.newValidationSignatureError();
            }
```

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/NestingAndCasingExpressionTest.java [hunk 5]
```diff
@@ -477,6 +477,11 @@ public void test_TO_EPOCH_MILLIS() {
         check(sql("TO_EPOCH_MILLIS(?) || TO_EPOCH_MILLIS(?)"), OFFSET_DATE_TIME_VAL, OFFSET_DATE_TIME_VAL);
     }
 
+    @Test
```
Comment (gierlachg) [inline hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/NestingAndCasingExpressionTest.java:480]: Move it next to `test_CONCAT`?

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        check(getConcatWsExpression("-", "1", "2"), "1-2");
+        check(getConcatWsExpression("-", "'1'", "2"), "1-2");
+    }
+
+    @Test
+    public void testNull() {
+        put("1");
+
+        //Null on separator => returns null
+        checkFail(getConcatWsExpression("null", false, "3", "2"));
```
Comment (gierlachg) [inline hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java:94]: Not sure if this should fail. Postgres simply returns `null` 🤔

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.sql.impl.type.QueryDataType.INT;
+import static com.hazelcast.sql.impl.type.QueryDataType.VARCHAR;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class ConcatWSFunctionIntegrationTest extends ExpressionTestSupport {
```
Comment (gierlachg) [inline hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java:41]: Would it make sense to add a test with all types?

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java (hunk 6 above)
Comment (yamanogluberk) [reply hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/string/ConcatWSFunctionIntegrationTest.java:94]: Yeah that was the intended behaviour, the `checkFail` method was working impaired it turns out, now its fixed

**Review by gierlachg — APPROVED** [review]

**Review by Fly-Style — APPROVED** [review]

**Review by hakanmemisoglu — APPROVED** [review]

**Review by viliam-durina — APPROVED** [review]

**Comment by gierlachg** [comment]
Worth mentioning new function in docs @JakeSCahill

**Review by viliam-durina — APPROVED** [review]

**Review by Fly-Style — APPROVED** [review]

**Comment by viliam-durina** [comment]
Congrats to your first PR! 😉

**Pull request merged**